include/fiction/algorithms/iter/gray_code_iterator.hpp

```diff
@@ -0,0 +1,295 @@
+
+        binary_to_gray();
+
+        return *this;
+    }
+#pragma GCC diagnostic pop
+    /**
+     * Assignment operator. Sets the current number to the given integer.
+     *
+     * @param m The number to set.
+     */
+    void operator=(const uint64_t m) noexcept
+    {
+        current_number = m;
+        binary_to_gray();
+    }
+    /**
+     * Equality operator. Compares the current number with the given integer.
+     *
+     * @param m Integer to compare with.
+     * @return `true` if the current number is equal to `m`, `false` otherwise.
+     */
+    [[nodiscard]] bool operator==(const uint64_t m) const noexcept
+    {
+        return current_number == m;
+    }
+    /**
+     * Inequality operator. Compares the current number with the given integer.
+     *
+     * @param m Integer to compare with.
+     * @return `true` if the current number is not equal to `m`, `false` otherwise.
+     */
+    [[nodiscard]] bool operator!=(const uint64_t m) const noexcept
+    {
+        return current_number != m;
+    }
+    /**
+     * Less-than operator. Compares the current number with the given integer.
+     *
+     * @param m Integer to compare with.
+     * @return `true` if the current number is less than `m`, `false` otherwise.
+     */
+    [[nodiscard]] bool operator<(const uint64_t m) const noexcept
+    {
+        return current_number < m;
+    }
+    /**
+     * Less-or-equal-than operator. Compares the current number with the given integer.
+     *
+     * @param m Integer to compare with.
+     * @return `true` if the current number is less than or equal to `m`, `false` otherwise.
+     */
+    [[nodiscard]] bool operator<=(const uint64_t m) const noexcept
+    {
+        return current_number <= m;
+    }
+    /**
+     * Greater-than operator. Compares the current number with the given integer.
+     *
+     * @param m Integer to compare with.
+     * @return `true` if the current number is greater than `m`, `false` otherwise.
+     */
+    [[nodiscard]] bool operator>(const uint64_t m) const noexcept
+    {
+        return current_number > m;
+    }
+    /**
+     * Greater-or-equal-than operator. Compares the current number with the given integer.
+     *
+     * @param m Integer to compare with.
+     * @return `true` if the current number is greater than or equal to `m`, `false` otherwise.
+     */
+    [[nodiscard]] bool operator>=(const uint64_t m) const noexcept
+    {
+        return current_number >= m;
+    }
+    /**
+     * Returns the start number, i.e., number for which the first Gray code is determined.
+     *
+     * @return Start number of the iteration.
+     */
+    [[nodiscard]] uint64_t get_start_number() const
+    {
+        return start_number;
+    }
+    /**
+     * Returns the current Gray code of the current iteration step.
+     *
+     * @return Current Gray code.
+     */
+    [[nodiscard]] uint64_t get_current_gray_code() const
+    {
+        return current_gray_code;
+    }
```

Comment:
Rather use `operator*`, otherwise, this class does not fulfill all requirements of an STL iterator.